### DISCUSSION TITLE
Override CopyPackageTranslationFiles Build Target

### DISCUSSION THREAD

**Discussion by jeffolmstead** [OrchardCMS/OrchardCore#9690] [Q&A] [answered]
@jtkech more than likely this is in your realm of expertise so thanks if you have any thoughts. I am trying to override the Build process to not have the target "CopyPackageTranslationFiles" run. In short, it slows down the buildout process when it has to check if the files localization files have changed (and right now I am only targeting one language). I am using a custom solution that references OrchardCore.Application.Cms.Targets in the project file. Therefore, from what I have read, in order to override the "CopyPackageTranslationFiles" target, I thought this would take care of it:

```
  <!-- CHANGE - START (needs to be after OrchardCore.Application.Cms.Targets or any project that references) -->
  <Target Name="CopyPackageTranslationFiles">
    <Message Text="Removing copy translation files target: $(MSBuildProjectName)" Importance="high" />
  </Target>
  <!-- CHANGE - END -->
``` 

But it doesn't and the translation files are still copied. Then I tried to see if I had any influence and added this:

```
 <!-- CHANGE - START (needs to be after OrchardCore.Application.Cms.Targets or any project that references) -->
  <Target Name="OverrideCopyPackageTranslationFiles" BeforeTargets="CopyPackageTranslationFiles">
    <Message Text="Removing copy translation files target: $(MSBuildProjectName)" Importance="high" />
    <PropertyGroup>
      <IsCrossTargetingBuild>true</IsCrossTargetingBuild>
    </PropertyGroup>
    <!--<Exec Condition="'$(IsCrossTargetingBuild)'=='true'" />-->
  </Target>
  <!-- CHANGE - END -->
```

This shows the Message text properly but of course, it doesn't remove the "CopyPackageTranslationFiles" from running. I messed around with trying to set the property so that the condition wouldn't fire, but currently I am stuck, nothing seems to prevent it from running. Look forward to any thoughts you might have and be willing to share.

**Comment by jtkech** [comment, marked as answer]
@jeffolmstead 

Yes, don't try to override properties as `IsCrossTargetingBuild`.

To override the target you need to also use the same `AfterTargets="Build"`.

Just tried by using the following **at the end** of the main app project file (seems that the last wins)

    <Target Name="CopyPackageTranslationFiles" AfterTargets="Build">
      <Message Text="For testing" Importance="high" />
    </Target>

It works but by still referencing OC as projects, i let you try it by referencing OC as packages ;)

If it don't work we may need to introduce a new msbuild property, e.g. `$(UpdatePackageTranslationFiles) = true / false`.

**Reply by jtkech** [reply]
Maybe another way is to create your own Cms.Targets meta package, just a copy paste of the original one, but in place of referencing `OrchardCore.Translations.All` just reference the translation packages that you need.

**Reply by jtkech** [reply]
Or create a target that runs before `CopyPackageTranslationFiles` and that clears the `@(PackageTranslationFiles)` collection that is used as a condition, going to sleep, will see tomorrow ;) https://github.com/OrchardCMS/OrchardCore/blob/c9e6a97d227ad5a32c107a98597da811125acf0d/src/OrchardCore/OrchardCore.Application.Cms.Core.Targets/OrchardCore.Application.Cms.Core.Targets.targets#L33

**Reply by jeffolmstead** [reply]
@jtkech thanks for getting back to me! I have a couple comments to your thoughts:

> To override the target you need to also use the same `AfterTargets="Build"`.
> 
> Just tried by using the following **at the end** of the main app project file (seems that the last wins)
> 
> ```
> <Target Name="CopyPackageTranslationFiles" AfterTargets="Build">
>   <Message Text="For testing" Importance="high" />
> </Target>
> ```
> 
> It works but by still referencing OC as projects, i let you try it by referencing OC as packages ;)

I have tried this approach and apparently with packages this does not work (at least I haven't been able to get it to win). 

> Maybe another way is to create your own Cms.Targets meta package, just a copy paste of the original one, but in place of referencing `OrchardCore.Translations.All` just reference the translation packages that you need.

I am sure this one would work, only negative would be managing this file whereas right now as it is changed in GitHub I just inherit any changes with the Package so if I go down this road, I would need to update it occassionally.

> Or create a target that runs before `CopyPackageTranslationFiles` and that clears the `@(PackageTranslationFiles)` collection that is used as a condition

Tried this but it too didn't seem to clear it in the Nuget package.

I stumbled onto something that does work. If I add in a file called "Directory.Build.targets" and put the above discussion into that file, it DOES override things and works perfectly. For completeness on anyone that stumbles upon this:

1. Create file called "Directory.Build.targets": 
![image](https://user-images.githubusercontent.com/1848585/122035859-e7a8e880-cda0-11eb-8979-69dd662a048e.png)
2. Paste this text into that file:
```
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Target Name="CopyPackageTranslationFiles" AfterTargets="Build">
    <Message Text="Removing the copying translation files: $(MSBuildProjectName)" Importance="high" />
  </Target>
</Project>
```

You don't need to add a reference in your project, it is automatically picked up. It is discussed [here](https://docs.microsoft.com/en-us/visualstudio/msbuild/customize-your-build)  Thanks @jtkech for the fantastic ideas, I have learned a lot about the build process.

**Reply by jtkech** [reply]
Cool 👍

Yes we also use `Directory.Build.targets` files, see under the parent folders of our modules / themes / OC libraries.

As mentioned, just tried the following that works

    <Target Name="Testing" BeforeTargets="CopyPackageTranslationFiles">
      <Message Text="Testing" Importance="high" />
      <ItemGroup>
        <PackageTranslationFiles Remove="@(PackageTranslationFiles)" />
      </ItemGroup>
    </Target>

You will still see the `Copying translation files` message but nothing wil be copied.

**Reply by jeffolmstead** [reply]
@jtkech so, in theory, this process could also be used by users who just want to have one or two translation language come over, correct? I am developing in English, but assuming someone wanted the French translations and only the French translations, the above process might allow for just copying over the translation files for French (instead of ALL the translation files). Mentioning it in case someone down the road stumbles upon this.

As a side note, this has dramatically improved my performance in changing code then debugging cycle.

**Reply by jtkech** [reply]
@jeffolmstead Yes exactly

E.g. if you only want the french translations, try the following


    <Target Name="Testing" BeforeTargets="CopyPackageTranslationFiles">
      <Message Text="Testing: %(PackageTranslationFiles.RecursiveDir)" Importance="high" />
      <ItemGroup>
        <FrenchTranslationFiles
          Include="@(PackageTranslationFiles)"
          Condition="'%(PackageTranslationFiles.RecursiveDir)' == 'fr\'" />
        <PackageTranslationFiles Remove="@(PackageTranslationFiles)" />
        <PackageTranslationFiles Include="@(FrenchTranslationFiles)" />
      </ItemGroup>
    </Target>
